utils/env_preferences/src/parse/posix.rs

```diff
@@ -0,0 +1,621 @@
+                .unwrap_or(src.len());
+
+            // Make sure this section is non-empty (more characters than just the delimiter)
+            if start_offset + 1 >= end_offset {
+                return Err(ParseError::EmptySection {
+                    offset: *start_offset,
+                });
+            }
+
+            // Write the section to the appropriate field
+            let section_value = Some(&src[start_offset + 1..end_offset]);
+            match delimiter {
+                Delimiter::Territory => locale.territory = section_value,
+                Delimiter::Codeset => locale.codeset = section_value,
+                Delimiter::Modifier => locale.modifier = section_value,
+            }
+        }
+
+        Ok(locale)
+    }
+
+    /// Attempt to convert a POSIX locale into a valid BCP-47 identifier.
+    ///
+    /// This is a best-effort conversion process, and there are valid
+    /// POSIX locales that will return an error or silently ignore data.
+    /// In particular, the codeset section is always ignored, and only some common modifiers are handled
+    /// (unknown modifiers will be silently ignored).
+    ///
+    /// If the "logging" feature is enabled, this ignored data will be logged as a warning.
+    pub fn try_convert_lossy(&self) -> Result<Locale, ConversionError> {
+        // Check if the language needs to be mapped to a BCP-47 language subtag
+        let language_code = POSIX_LANGUAGE_NAMES
+            .binary_search_by_key(&self.language, |(language, _code)| language)
+            // `binary_search_by_key()` guarantees a valid index
+            .map(|index| POSIX_LANGUAGE_NAMES[index].1);
+
+        let mut language = match language_code {
+            // Found a matching language code
+            Ok(language) => language,
+            // Parse the language directly
+            Err(_) => match Language::try_from_str(self.language) {
+                Ok(language) => language,
+                Err(_) => {
+                    return Err(ConversionError::InvalidLanguage {
+                        start_offset: 0,
+                        end_offset: self.language.len(),
+                    })
+                }
+            },
+        };
+
+        let region = if let Some(territory) = self.territory {
+            match Region::try_from_str(territory) {
+                Ok(region) => Some(region),
+                Err(_) => {
+                    return Err(ConversionError::InvalidRegion {
+                        // Add 1 to skip the delimiter
+                        start_offset: self.language.len() + 1,
+                        // Add 1 to include the final character
+                        end_offset: self.language.len() + territory.len() + 1,
+                    });
+                }
+            }
+        } else {
+            None
+        };
+
+        if let Some(codeset) = self.codeset {
+            icu_provider::log::warn!("Ignoring codeset: `{codeset}`");
+        }
```

Comment:
`question`: I'm confused about the use of provider logging here. This is a library, not a program. My understanding is that we don't do UX (logging is a form of ux) in libraries. What I'd expect is that if you need this you could add information to the return value, or allow for outparam that you populate with such information as "omitted fields".

But frankly, I don't think this is needed at all. You're trying to construct a Unicode BCP47 Locale from POSIX Locale. It's lossy, best attempt effort. No need to document what was omitted I think.